refactor: Pass WalletModel object to the WalletView constructor

hebasto opened the pull request:
An instance of the `WalletView` class without the `walletModel` data member being set is invalid. So, it is better to set it in the constructor.

Establishing one more `WalletView` class's invariant in constructor:
- allows to drop all of checks of the`walletModel` in member functions
- makes reasoning about the code that uses instances of the `WalletView` class  easier

Possible follow ups could extend this approach to other classes, e.g., `OverviewPage`, `TransactionView`, `ReceiveCoinsDialog`, `SendCoinsDialog`, `AddressBookPage`.

Talkless commented:
Concept ACK

Talkless commented on src/qt/walletview.cpp:
```diff
@@ -100,49 +127,15 @@ void WalletView::setClientModel(ClientModel *_clientModel)
-        updateEncryptionStatus();
-
-        // update HD status
-        Q_EMIT hdEnabledStatusChanged();
```
I don't see this signal being emmited in new code. It's because initialization in the constructor, as there's no "change" per se?

Talkless commented on src/qt/walletview.h:
```diff
@@ -35,26 +35,26 @@ class WalletView : public QStackedWidget
-        functionality.
-    */
-    void setWalletModel(WalletModel *walletModel);
+    WalletModel* getWalletModel() const { return walletModel; }
```
Maybe worth adding `noexcept` and `[[nodiscard]]` too? Or at least just `noexcept`, as it's not "heavyweight" operation to ask for "nodsicardness", but still...

Talkless replied:
Also forgot to add, `WalletModel *walletModel;` pointer itself now can be `const`.

hebasto replied:
> Also forgot to add, `WalletModel *walletModel;` pointer itself now can be `const`.

```
In file included from qt/moc_walletview.cpp:9:
./qt/walletview.h:45:50: error: cannot initialize return object of type 'WalletModel *' with an lvalue of type 'const WalletModel *const'
    WalletModel* getWalletModel() const { return walletModel; }
                                                 ^~~~~~~~~~~
1 error generated
```

Talkless replied:
Not `const WalletModel *` but `WalletModel *const walletModel`. Pointer itself `const`, not the object pointed to.

Talkless reviewed (approved):
utACK bb56486a170aacb355f4a973f0cd40ab3918a0cd

hebasto commented:
@Talkless 

> I don't see this signal being emmited in new code. It's because initialization in the constructor, as there's no "change" per se?

#403 could make things simpler.

So labeling this PR as a draft for now, and begging to review #403 first.

promag commented on src/qt/bitcoingui.cpp:
```diff
@@ -676,7 +676,7 @@ void BitcoinGUI::addWallet(WalletModel* walletModel)
 
-    WalletView* wallet_view = new WalletView(platformStyle, walletFrame);
+    WalletView* wallet_view = new WalletView(walletModel, platformStyle, walletFrame);
     if (!walletFrame->addWallet(walletModel, wallet_view)) return;
```
Could a commit to simplify like
```diff
-bool WalletFrame::addWallet(WalletModel* walletModel, WalletView* walletView)
+bool WalletFrame::addView(WalletView* walletView)
```

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin-core/gui/pull/398#issuecomment-912071917).

hebasto commented:
Updated 92ddc02a16a74e10f24190929f05e2dcf2b55871 -> d319c4dae9ed7d59d71b926e677707fce4194d0c ([pr398.02](https://github.com/hebasto/gui/commits/pr398.02) -> [pr398.03](https://github.com/hebasto/gui/commits/pr398.03), [diff](https://github.com/hebasto/gui/compare/pr398.02..pr398.03)):

- addressed @promag's [comment](https://github.com/bitcoin-core/gui/pull/398#discussion_r701421756)

ShaMan239 reviewed (commented):
Code review ACK d319c4dae9ed7d59d71b926e677707fce4194d0c

These changes seem to simplify the code a great bit. Thanks for this amazing work!

I have compiled some notes regarding the changes made in each of the commits in this PR. However, @hebasto has very clearly mentioned the details in the commit message of each commit. I still think my notes might be of some help to fellow reviewers.

NOTES:
- **qt, refactor: Pass WalletModel object to WalletView constructor**
    - The process of creating `walletmodel` has been moved under the constructor of the `WalletView` class.
    - The redundant function of creating a `walletmodel` has been removed.
    - `walletview.h` file has been appropriately modified to account for all the changes made in the `walletview.cpp` file.
- **qt, refactor: Drop redundant checks of walletModel**
    - As the name suggests, all the redundant checks of `walletModel` in the `walletview.cpp` file have been removed, as the `walletmodel` is now constructed during the construction of the `walletview` object itself.
- **qt, refactor: Declare getWalletModel with const and noexcept qualifiers** 
    - This is done to ensure that the function is not changing a member variable’s value. And the given function will not throw an exception (since the `walletmodel` is generated with the constructor itself).
- **qt, refactor: Replace WalletFrame::addWallet with WalletFrame::addView**
    - Modified the `addWallet` function such that it now doesn’t need to take `walletmodel` as an argument. And also removed the need to check for the existence of a `walletmodel` inside this function.

If some of my observation is erroneous or lacking thorough observation, please do inform me. 
<p>

promag reviewed (commented):
Code review ACK d319c4dae9ed7d59d71b926e677707fce4194d0c.

hebasto commented:
@Talkless @jarolrod 

Do you mind having another look into this PR?

jarolrod reviewed (commented):
ACK d319c4dae9ed7d59d71b926e677707fce4194d0c

I did light testing with different wallets and switching between them. I also ran the qt tests locally. Performed identical testing steps on macOS and on Ubuntu.

This is a nice simplification of the WalletModel/WalletView logic. It ties up nicely at the end with the last commit where we now only need to pass the WalletView object; not both a WalletModel and WalletView object.

@ShaMan239 your notes on the PR are excellent and I looked at them before beginning my review. Thanks for posting them!
